nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/TestGetOrCreateDocument.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * (C) Copyright 2017 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Funsho David
+ *
+ */
+
+package org.nuxeo.ecm.core;
+
+import org.apache.commons.lang3.mutable.MutableObject;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.nuxeo.ecm.core.api.ConcurrentUpdateException;
+import org.nuxeo.ecm.core.api.CoreInstance;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.DocumentRef;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.kv.KeyValueService;
+import org.nuxeo.ecm.core.kv.KeyValueStore;
+import org.nuxeo.ecm.core.test.CoreFeature;
+import org.nuxeo.ecm.core.test.annotations.Granularity;
+import org.nuxeo.ecm.core.test.annotations.RepositoryConfig;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.test.runner.Deploy;
+import org.nuxeo.runtime.test.runner.Features;
+import org.nuxeo.runtime.test.runner.FeaturesRunner;
+import org.nuxeo.runtime.transaction.TransactionHelper;
+
+import javax.inject.Inject;
+import java.io.Serializable;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.nuxeo.ecm.core.kv.KeyValueConstants.DOCUMENT_LOCK;
+
+/**
+ * @since 9.3
+ */
+@RunWith(FeaturesRunner.class)
+@Features(CoreFeature.class)
+@RepositoryConfig(cleanup = Granularity.METHOD)
+@Deploy("org.nuxeo.ecm.core.kv")
+public class TestGetOrCreateDocument {
+
+    @Inject
+    protected CoreFeature coreFeature;
+
+    @Inject
+    protected CoreSession session;
+
+    protected DocumentModel getOrCreateDocument(CoreSession session, DocumentModel docModel) {
+        return CoreInstance.doAtomically(session, docModel, obj -> computeKey((DocumentModel) obj),
+                (CoreSession s, Serializable obj) -> {
+                    try {
+                        return doGetOrCreateDocument(s, (DocumentModel) obj);
+                    } catch (InterruptedException | ConcurrentUpdateException e) {
```

Comment:
Always catch `InterruptedException` in its own block and do `Thread.currentThread().interrupt(); // restore interrupt` before rethrowing.